File: ui/app/src/lib/api/permissions.ts
import { API_BASE, ApiError, apiFetch, readErrorBody } from './core.ts';

/**
 * REST client for per-vault persisted agent "Always Allow" grants (issue #189).
 *
 * Mirrors `transcripts.ts`: persistence is frontend-orchestrated. The chat
 * store fetches a vault's grants at session start (to pre-seed the ACP session
 * so granted tools never re-prompt) and POSTs a new grant when the user picks
 * "Always Allow". The daemon owns the durable store; the Rust agent layer stays
 * HTTP-free.
 */

function base(vault: string): string {
	return `${API_BASE}/api/v/${encodeURIComponent(vault)}/agent/permissions`;
}

async function fail(res: Response, fallback: string): Promise<never> {
	const { code, message } = await readErrorBody(res);
	throw new ApiError(message ?? `${fallback}: ${res.status}`, res.status, code);
}

/**
 * A 404 on the permissions collection route means the daemon predates the
 * agent-permission API — a version-skew, not a missing grant. Surface a clear,
 * actionable message instead of a bare "404" (matching `transcripts.ts`).
 */
function outdatedServerError(): ApiError {
	return new ApiError(
		"Persisted agent permissions aren't available: this Notesmith server is too old (missing the agent permission API). Update the daemon to the latest version.",
		404,
		'permission_api_unavailable'
	);
}

/** List the tool names the user has granted "Always Allow" for `vault`. */
export async function listGrants(vault: string): Promise<string[]> {
	const res = await apiFetch(base(vault));
	if (res.status === 404) throw outdatedServerError();
	if (!res.ok) return fail(res, 'Failed to list permission grants');
	return res.json();
}

/** Persist an "Always Allow" grant for `tool` in `vault`. */
export async function grant(vault: string, tool: string): Promise<void> {
	const res = await apiFetch(base(vault), {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ tool })
	});
	if (res.status === 404) throw outdatedServerError();
	if (!res.ok) return fail(res, 'Failed to grant permission');
}

/** Revoke a persisted grant for `tool` in `vault`. */
export async function revoke(vault: string, tool: string): Promise<void> {
	const res = await apiFetch(`${base(vault)}/${encodeURIComponent(tool)}`, {
		method: 'DELETE'
	});
	if (!res.ok && res.status !== 404) return fail(res, 'Failed to revoke permission');
}

File: crates/notesmith-http/src/routes/permissions.rs
//! HTTP endpoints for per-vault persisted agent write grants (issue #189).
//!
//! These expose the daemon-owned
//! [`PermissionGrantStore`](notesmith_permission::PermissionGrantStore) so the
//! desktop chat UI can fetch a vault's "Always Allow" grants at session start
//! (to pre-seed the ACP session and never re-prompt) and persist a new grant
//! when the user picks "Always Allow". Persistence is frontend-orchestrated:
//! the Rust agent/ACP layer stays HTTP-free and only consumes the seed list.
//!
//! All routes are vault-scoped (`/api/v/{vault}/agent/permissions`); the store
//! keys every operation by vault, so one vault's grants are never visible under
//! another. Bodies are validated and malformed input yields a structured 4xx
//! (never a 500), matching `routes::transcripts`.

use axum::{
    Json,
    extract::{Path, State},
    http::StatusCode,
};
use notesmith_permission::PermissionError;
use serde::Deserialize;
use serde_json::{Value, json};

use crate::server::SharedAppState;

type ApiError = (StatusCode, Json<Value>);

fn error(status: StatusCode, message: impl std::fmt::Display) -> ApiError {
    (status, Json(json!({ "error": message.to_string() })))
}

fn store_error(err: PermissionError) -> ApiError {
    tracing::error!(error = %err, "permission store error");
    error(StatusCode::INTERNAL_SERVER_ERROR, "permission store error")
}

/// Clone the shared permission store out of `AppState`, releasing the app lock
/// before doing (synchronous) database work.
async fn permissions(
    state: &SharedAppState,
) -> std::sync::Arc<notesmith_permission::PermissionGrantStore> {
    state.read().await.permissions.clone()
}

#[derive(Debug, Deserialize)]
pub struct GrantRequest {
    pub tool: String,
}

/// `GET /api/v/{vault}/agent/permissions` — list a vault's persisted grants.
pub async fn list_grants(
    State(state): State<SharedAppState>,
    Path(vault): Path<String>,
) -> Result<Json<Vec<String>>, ApiError> {
    let store = permissions(&state).await;
    let granted = store.list_granted(&vault).map_err(store_error)?;
    Ok(Json(granted))
}

/// `POST /api/v/{vault}/agent/permissions` — persist an "Always Allow" grant.
pub async fn grant_permission(
    State(state): State<SharedAppState>,
    Path(vault): Path<String>,
    Json(req): Json<GrantRequest>,
) -> Result<StatusCode, ApiError> {
    if req.tool.trim().is_empty() {
        return Err(error(StatusCode::BAD_REQUEST, "tool must not be empty"));
    }
    let store = permissions(&state).await;
    store.grant(&vault, &req.tool).map_err(store_error)?;
    Ok(StatusCode::NO_CONTENT)
}

/// `DELETE /api/v/{vault}/agent/permissions/{tool}` — revoke a persisted grant.
pub async fn revoke_permission(
    State(state): State<SharedAppState>,
    Path((vault, tool)): Path<(String, String)>,
) -> Result<StatusCode, ApiError> {
    let store = permissions(&state).await;
    store.revoke(&vault, &tool).map_err(store_error)?;
    // Revoking an absent grant is idempotent; 204 either way.
    Ok(StatusCode::NO_CONTENT)
}
